pkg/consolegraphql/server/session.go

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright 2020, EnMasse authors.
+ * License: Apache License 2.0 (see the file LICENSE or http://apache.org/licenses/LICENSE-2.0.html).
+ *
+ */
+
+package server
+
+import (
+	"github.com/alexedwards/scs/v2"
+	"github.com/enmasseproject/enmasse/pkg/consolegraphql/server/session"
+	"github.com/prometheus/client_golang/prometheus"
+	"sync/atomic"
+	"time"
+)
+
+type sessionCountingListener struct {
+	sessionCount      int32
+	sessionPresent    func()
+	noSessionsPresent func()
+	metric            prometheus.Gauge
+}
+
+func (l *sessionCountingListener) Added(string) {
+	curr := atomic.AddInt32(&l.sessionCount, 1)
+	if curr == 1 {
```

Comment:
Is the assumption that Added is called only once for a session? If not, why is sessionPresent() only called if == 1 and not >= 1? Maybe add some comment to clarify how this works.

Reply:
`Added` is called each time a new session is started.  `Removed` is called each time a session is removed (or is expired).   The watcher manager needs to be notified when the session store transitions from having no sessions to having the first, and the reverse, from having one to having none so that it can begin and end the watchers.    At the moment the watcher manager is not idempotent, so it is important that the begin and end messages are received exactly once.  To achieve this I used to atomic int in the sessions listeners and trigger when the atomic ints value becomes 0 or 1 accordingly.

Reply:
Thanks for the explanation! Maybe put it into a comment?

Reply:
May be making watcher matcher idempotent would make the code clearer and address your comment?